feature/ added form combo box

DexinJ opened the pull request:
## Describe your changes
added form combo box, installed insider version of the headless ui package for combobox options to be displayed when focused.
## Issue ticket number and link
[slop-13](https://tripleten-apiary.atlassian.net/jira/software/projects/SLOP/boards/2?selectedIssue=SLOP-13)
## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] If there are changes to components, I have added / updated automated tests
- [x] I have made any necessary updates to Storybook to accompany these changes
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

SamPage4427 commented:
Looks good so far, there are a few issues with the combo box that should probably be addressed at this stage.

1. The combo box is going to be multiuse, so using item.name is probably not the best way to handle setting the values. For example the create slop fest and edit slop fest modals will use the combo box and they would be using item.username and not item.name. For those modals the combo box will not work but for a keyword combo box where the keyword is most likely to have a item.name attribute it would work perfectly fine.
2. When testing the combo box, I noticed when there are multiple values selected and you click the x to delete one of them, all of the values disappear.

DexinJ commented:
> 1. The combo box is going to be multiuse, so using item.name is probably not the best way to handle setting the values. For example the create slop fest and edit slop fest modals will use the combo box and they would be using item.username and not item.name. For those modals the combo box will not work but for a keyword combo box where the keyword is most likely to have a item.name attribute it would work perfectly fine.
I have added a way for people to define the key name for id and name when passing object into the combo box
> 2. When testing the combo box, I noticed when there are multiple values selected and you click the x to delete one of them, all of the values disappear.
I'm not sure what's causing this as it works fine on my end, but I added another condition where it will keep all selected options with different names as well.

jduncan017 commented on src/components/form/form.jsx:
```diff
@@ -128,3 +132,74 @@ Form.Feedback = ({ className, message }) => {
+            <Combobox.Option
+              key={item[idKey]}
+              value={item}
+              className="font-normal py-3 px-4 border-solid rounded-none border border-black"
```
Adding the tailwind classes cursor-pointer and hover:bg-(some color) will help dramatical with the user experience here

jduncan017 replied:
Even though the options are displayed in the input of the combobox, adding a highlight to options that have been selected in the dropdown would be helpful as well. This way the user can see which options they've already selected.

jduncan017 commented on src/components/form/form.jsx:
```diff
@@ -128,3 +132,74 @@ Form.Feedback = ({ className, message }) => {
+            <img src={down} className="h-2.5 w-2.5" />
+          </Combobox.Button>
+        </div>
+        <Combobox.Options>
```
When there are a lot of options, the combobox options just continue to push the page down. Without solving for this directly I'd have to play with solutions but I'd image you would need to display the options within a div that is positioned absolutely over the rest of the page. Additionally, this div should have a max height and overscroll-auto so that the user can scroll through the options instead having one long list.

jduncan017 commented on src/components/form/form.jsx:
```diff
@@ -128,3 +132,74 @@ Form.Feedback = ({ className, message }) => {
+                <span>{item[nameKey]}</span>
+                <button
+                  type="click"
+                  onClick={() => {
```
I think this onClick handler would be better to add to the entire input, so that the user can see the items when they interact with the input.

DexinJ replied:
I'm gonna need a little clarification here. If I'm not mistaken, this is the button that deselects the item from the selected list.
So you want the user to be able to deselect by clicking on the entire thing and not just the button?

jduncan017 replied:
I'm responding on mobile and can't see all of the code as this comment box is covering it. If this is the button to deselect an item then I was mistaken.

What I meant was that the options should display when the user clicks on the input of the combo box and not just when they click on the drop down button on the right. I believe this is how most combo boxes operate across the web and would be more intuitive.

jduncan017 commented on src/components/form/form.jsx:
```diff
@@ -128,3 +132,74 @@ Form.Feedback = ({ className, message }) => {
+        </div>
+        <Combobox.Options>
+          {filteredList.map((item) => (
+            <Combobox.Option
```
Keyboard navigation of the items in the dropdown would be a nice touch, allowing the user to navigate between items in the dropdown with arrowkeys

jduncan017 commented on src/components/form/form.jsx:
```diff
@@ -128,3 +132,74 @@ Form.Feedback = ({ className, message }) => {
+              >
+                <span>{item[nameKey]}</span>
+                <button
+                  type="click"
```
type should = "button" here

jduncan017 commented on src/components/form/form.jsx:
```diff
@@ -128,3 +132,74 @@ Form.Feedback = ({ className, message }) => {
+        });
+  return (
+    <div className={`flex font-bold font-arial flex-col py-3 ${className}`}>
+      <label htmlFor={id} className={`mb-1.5 text-lg text-start `}>
```
The label has an htmlFor declared but it's never used to tie it to the combobox.input

DexinJ replied:
everything should be fixed now

jduncan017 reviewed (changes requested):
This seems like a pretty challenging component to create. Great job it's going to be really useful for the site. Please note the comments there a few places I think it could be improved.

jduncan017 reviewed (approved):
Looks great man thank you!

jahorwitz commented:
Great job! Love all the PR comments back and forth 👍
